crates/utils/src/http_client.rs

```diff
@@ -0,0 +1,845 @@
+        });
+
+        // Test malformed response
+        let mock_malformed = mock_server.mock(|when, then| {
+            when.method("GET").path("/malformed");
+
+            then.status(200).header("content-type", "application/json").body("invalid json");
+        });
+
+        let client = HttpClient::builder(&mock_server.base_url()).build().unwrap();
+
+        // Test 404 response
+        let response = client.request().method(Method::GET).path("/not-found").send().await.unwrap();
+        assert_eq!(response.status(), 404);
+        mock_404.assert();
+
+        // Test 500 response
+        let response = client.request().method(Method::GET).path("/server-error").send().await.unwrap();
+        assert_eq!(response.status(), 500);
+        mock_500.assert();
+
+        // Test malformed response
+        let response = client.request().method(Method::GET).path("/malformed").send().await.unwrap();
+        assert_eq!(response.status(), 200);
+        let text = response.text().await.unwrap();
+        assert_eq!(text, "invalid json");
+        mock_malformed.assert();
+    }
+
+    /// # Error Handling Tests
+
+    /// Tests TLS certificate and identity handling:
+    /// - Valid certificate addition
+    /// - Invalid certificate handling
+    /// - Identity verification
+    #[test]
+    fn test_certificate_handling() {
+        let mock_server = httpmock::MockServer::start();
+
+        // Load variables from .env.test
+        dotenv::from_filename(".env.test").ok();
+
+        // Getting the cert files from the .env.test and then decoding it from base64
+        let cert = general_purpose::STANDARD
+            .decode(std::env::var("SHARP_USER_CRT").unwrap())
+            .expect("Failed to decode certificate");
+        let key = general_purpose::STANDARD
+            .decode(std::env::var("SHARP_USER_KEY").unwrap())
+            .expect("Failed to decode sharp user key");
+        let server_cert = general_purpose::STANDARD
+            .decode(std::env::var("SHARP_SERVER_CRT").unwrap())
+            .expect("Failed to decode sharp server certificate");
+
+        let identity =
+            Identity::from_pkcs8_pem(&cert, &key).expect("Failed to build the identity from certificate and key");
+        let certificate = Certificate::from_pem(server_cert.as_slice()).expect("Failed to add root certificate");
+
+        let client = HttpClient::builder(&mock_server.base_url())
+            .identity(identity)
+            .add_root_certificate(certificate)
+            .build()
+            .unwrap();
+
+        // Since we can't check the certificates directly, we'll just verify the client was built
+        assert_eq!(client.base_url.as_str(), mock_server.base_url() + "/");
+    }
+
+    /// # Edge Cases
+
+    /// Tests behavior with empty or problematic paths:
+    /// - Empty path segments
+    /// - Multiple slashes
+    /// - Unicode paths
+    #[test]
+    fn test_empty_path_handling() {
```

Comment:
let's have different cases for the same test case or in the same fn with multiple lines

Reply:
refactored